src/main.rs

```diff
@@ -439,7 +440,38 @@ fn main() -> anyhow::Result<()> {
         );
         handle.join().expect("thread is not panicking")?;
     }
-    drop(toc_arc);
+    drop(search_runtime);
+    drop(update_runtime);
+    drop(general_runtime);
     drop(settings);
+    drop(wait_unwrap(toc_arc));
     Ok(())
 }
+
+/// Wait for the current thread to become a single referent of an [`Arc`] and returns it.
+///
+/// This methods move a value out from an [`Arc`] when there is only one reference left
+/// and it is safe to do so.
+///
+/// **CAUTION**: this method may create live lock if parallel threads are not going to finish
+/// and decrement [`Arc`] counter. Caller should initiate shutdown of all the threads holding the
+/// [`Arc`] before calling this method.
+fn wait_unwrap<T>(mut input: Arc<T>) -> T {
+    let start_time = Instant::now();
+    let mut notified = false;
+    loop {
+        if !notified && start_time.elapsed() > Duration::from_secs(10) {
+            log::warn!("Waiting for Arc to be dropped...");
+            notified = true;
+        }
```

Comment:
let's return error instead of potentially infinite loop. Infinite loops are _bad_

Reply:
Ok, no problem.

Just to make sure we on the same page. The reason we are blocking main is that there is still background activity on the RocksDB side. The probable outcome of early return is segfault – the problem we are trying to solve in the first place. I will increase timeout to around 30 seconds to minimize probability of such a scenario.

Reply:
Ok, no problem. But just to make sure we're on the same page. The reason we are blocking the main is that there is still background activity on the RocksDB side. The probable outcome of an early return is SIGSEGV – the problem we are trying to solve in the first place.

I will increase the timeout to 30 seconds to minimize the probability of such a scenario if you don't mind.